src/content/docs/en/reference/integrations-reference.mdx

```diff
@@ -613,6 +613,121 @@ A list of all generated pages. It is an object with one property.
+      'myLib:eventHappened': (your: string, parameters: number) => Promise<void>;
+    }
+  }
+}
+```
+
+#### Calling custom hooks
+
+Installed integrations can be accessed from the `astro:config:setup` and `astro:config:done` hooks. Loop over these to call your custom hook on each if available:
+
+```ts {7-9}
+// your-integration/index.ts
+export default (): AstroIntegration => {
+  return {
+    name: 'your-integration',
+    hooks: {
+      'astro:config:setup': async ({ config }) => {
+        for (const integration of config.integrations) {
+          await integration.hooks['myLib:eventHappened'].?('your values', 123);
+        }
+      },
+    }
+  }
+}
+```
+
+To call custom hooks outside of `astro:config:setup` and `astro:config:done`, store the integrations array and use it later. `astro:config:done` should be preferred to work with integrations that get added by other integrations during `astro:config:setup`.
+
+```ts {3,9-10,13-16}
+// your-integration/index.ts
+export default (): AstroIntegration => {
+  let integrations: AstroIntegration[] = [];
+
+  return {
+    name: 'your-integration',
+    hooks: {
+      'astro:config:done': async ({ config }) => {
+        // Store the integrations when available
+        integrations = config.integrations;
+      },
+      'astro:build:done': async () => {
+        // Loop over the stored integrations in a later hook
+        for (const integration of config.integrations) {
+          await integration.hooks['myLib:eventHappened'].?('your values', 123);
+        }
+      },
+    }
+  }
+}
+```
+
+#### Using a custom hook from another integration
+
+An integration that uses your custom hook will need to import your types file. It can then provide a value for your hook in `hooks` with type checking and editor hints.
+
+```ts {6-8}
+// other-integration/index.ts
+export default (): AstroIntegration => {
+  return {
+    name: 'other-integration',
+    hooks: {
+      'myLib:eventHappened': async (your, values) => {
+        // ...
+      },
+    }
+  }
+}
+```
+
+Make sure that your type augmentation can be loaded by other integrations; otherwise, their hook implementations won't be properly typed. The easiest way to ensure this is to either have that declaration in the same file that exports your integration or import it from that file:
+
+```ts {2-3} {5-12}
+// your-integration/index.ts
+// Import the types
+import './types.ts';
+
+// Or declare it in the same file
+declare global {
+  namespace Astro {
+    interface IntegrationHooks {
+      'myLib:eventHappened': (your: string, parameters: number) => Promise<void>;
+    }
+  }
+}
+
+export default (): AstroIntegration => {
+  return {
+    name: 'your-integration',
+    hooks: { /* ... */ }
+  }
+}
+```
+
+In either case, other integrations can load your types by importing your integration or declaring it as a reference:
+
+```ts
+// other-integration/index.ts
+import yourIntegration from 'your-integration';
+// or
```

Comment:
Pick one!

Reply:
Contrary to the one above, these two options are not equal.

If you want to integrate with another integration _if_ it is there, but don't want to require your users to install them if they don't want to, you need to use the `/// <directive .../>` option.
If you are going to add the integration yourself if it is not already present in the project then you can just import it.

JS ecosystem is quite bloated with dependencies that are not used at all at runtime because people just go for the import 😞

Reply:
Ah, cool. (I mean, not cool, but good to know)...

In this case, I'd like to add that extra knowledge here then! I know I'm asking stupid questions that maybe integration authors already know, but to let you know where I'm coming from.. historically I haven't pushed back too much on any of these API reference docs since they are a more advanced use case and many people reading them will probably already have this context or understand. But, I'm trying to bring more explicitness to them as we add new content, and the best way for me to know that they really do make sense to any audience is if they make at least a bit of sense to me! (A tough ask, I know.)

So in a case like this, where someone has two choices (and they actually do different things), I am leaning towards explaining what these different options actually do. One option could be two tiny (whatever appropriate level) sub headings, e.g.

> ####### Without installing the other integration as a dependency
> ####### Installing the other integration as a user dependency